ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp

```diff
@@ -8982,18 +8982,56 @@ bool GDALGeoPackageDataset::AddRelationship(std::unique_ptr<GDALRelationship>&&
             osMappingTableName.Printf( "%s_%s_%d", osLeftTableName.c_str(), osRightTableName.c_str(), nIndex );
         }
 
+        // determine whether base/related keys are unique
+        bool bBaseKeyIsUnique = false;
+        {
+            const std::set<std::string> uniqueBaseFieldsUC = SQLGetUniqueFieldUCConstraints(GetDB(), osLeftTableName.c_str());
+            if ( uniqueBaseFieldsUC.find( CPLString( aosLeftTableFields[0] ).toupper() ) != uniqueBaseFieldsUC.end() )
+            {
+                bBaseKeyIsUnique = true;
+            }
+        }
+        bool bRelatedKeyIsUnique = false;
+        {
+            const std::set<std::string> uniqueRelatedFieldsUC = SQLGetUniqueFieldUCConstraints(GetDB(), osRightTableName.c_str());
+            if ( uniqueRelatedFieldsUC.find( CPLString( aosRightTableFields[0] ).toupper() ) != uniqueRelatedFieldsUC.end() )
+            {
+                bRelatedKeyIsUnique = true;
+            }
+        }
+
         // create mapping table
-        // note ideally we'd be setting foreign key constraints on this mapping table too, BUT this will cause sqlite
-        // foreign key mismatches if the base/related table fields aren't primary keys in their tables.
-        // "CONSTRAINT [fk_base_id_%q] FOREIGN KEY (base_id) REFERENCES \"%w\"(\"%w\") ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED, "
-        // "CONSTRAINT [fk_related_id_%q] FOREIGN KEY (related_id) REFERENCES \"%w\"(\"%w\") ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED"
+
+        std::string osBaseIdDefinition = "base_id INTEGER";
+        if ( bBaseKeyIsUnique )
+        {
+            char* pszSQL = sqlite3_mprintf(" CONSTRAINT [fk_base_id_%q] REFERENCES \"%w\"(\"%w\") ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED",
```

Comment:
does [foo_%q] work fine when there are single, double quote, space, [ or ] character in osMappingTableName ?

Reply:
I've reworked this, and added new tests to ensure everything works nicely with complex field/table names